atmost/atmost_driver.man
.TH ATMOST 1 "2019-08-06" "Limits concurrent execution of executable."
.SH "NAME"
atmost_driver \- driver server for the atmost command.
.SH "SYNOPSIS"
.SY atmost_driver
.OP \-s|\-\-socket socket\-path
.OP \-c|\-\-callbacks file\-path
.OP \-u|\-\-update interval
.YS
.
.SY atmost_driver
.B \-h
.SY atmost_driver
.B --help
.YS
.SH "DESCRIPTION"
\fBatmost_socket\fR acts as a server listening on a local Unix-domain socket
(address specified by the \fB-s\fR command-line argument), accepting connections
and handling requests requests from instances of the \fBatmost\fR command
and determining when the executable handled by \fBatmost\fB shall be executed.
.PP
The request from \fBatmost\fR takes the form of a JSON object with at least
the following mandatory attributes:
.IP \(dqargs\(dq
a JSON list of strings containing the command-line arguments of the
\fIexecutable\fB managed by \fBatmost\fR as they will be passed to the
\fBexec\fR system call.
.IP \(dqpid\(dq
the integer process identifier of the atmost client process handling the
\fIexecutable\fR.
.PP
The response is a simple string with the following format \(dqOK-GO:%d\(dq,
where \fB%d\fR is substituted with the process id of the receiving process.

.SH "OPTIONS"
.SS "General options"
.TP
\fB\-h\fR, \fB\-\-help\fR
Show a help screen and exit.
.TP
\fB\-s\fR \fIsocket-path\fR, \fB\-\-socket\fR \fIsocket-path\fR
Unix domain socket path on which the driver accepts connections from instances
of \fBatmost\fR. The \fBatmost\fR process is blocked until it receives a response
from the driver.
If this option is not specified, the default is \fI/tmp/atmost.socket\fB.
.TP
\fB\-u\fR \fIinterval\fR, \fB\-\-update\fR \fIinterval\fR
The atmost driver  updates the internal information tracking the process state
whenever new requests on the server socket arrive and also periodically.
This option changes the periodic update interval. The \fIinterval\fR argument
is a positive floating point value in seconds.
If this option is not specified, the default is 1 second.
.SS "Driver callbacks"
.PP
In order to manage and control the execution start time of the executables
handled by the \fBatmost\fR utility the driver uses a set of callback functions
which are imported by the driver from a user-specified Python script and invoked
at particular events during the execution of the driver.
.IP "def load_callback_data():"
This function is called when the driver is initializing and it gives the callback
script a chance to do global initialization and to load resources needed during
the whole execution time of \fBatmost_driver\fR.
The function may return a value, which is then stored within the driver and
passed as the argument to the other callbacks.
.IP "def save_callback_data(callback_data):"
This function is called when the driver is finalizing and it allows the callback
script to save any global state that should remain persistent between invocations
of \fBatmost_driver\fR. The argument \fIcallback_data\fB is the same object
that the \fBload_callback_data\fR function returned.
.IP "def process_initialized(callback_data, proc):"
This function is called after an instance of \fBatmost\fR connects to the driver
and providers the necessary information about the managed and controlled
\fIexecutable\fB. The \fIcallback_data\fR parameter is the same object that
was returned by the \fBload_callback_data\fR function. The \fIproc\fR argument
is a class providing information about the managed executable, like its
command-line and about the process in which it will be executed.
.IP "def let_process_go(context, procs):"
This function is called repeatedly to determine if a waiting process can start
executing its managed \fIexecutable\fB.
The \fIcallback_data\fR parameter is the same object that
was returned by the \fBload_callback_data\fR function.
The \fIprocs\fR argument is a class providing information about all currently
listed managed processes. The processes are divided into three groups:
.IP \[bu] 2
\fIactive\fR: list of processes already running the managed \fIexecutable\fB
and consuming system resources.
.IP \[bu] 2
\fIcurrent\fR: the process for which we are determining if it should start execution.
.IP \[bu] 2
\fIwaiting\fR: list of processes prepared to be executed, but which were not yet
let go.  This list does not contain the \fIcurrent\fR process.
.PP
The \fIprocs\fR object provides functions active(), current() and waiting()
returning the groups of processes described above.
.PP
The function shall return \fITrue\fR if the managed executable can start execution,
or return \fIFalse\fR otherwise.
.IP "def process_finished(callback_data, proc):"
This function is called after a managed \fBatmost\fR process has finished the
execution if the managed \fIexecutable\fB.
The \fIcallback_data\fR parameter is the same object that
was returned by the \fBload_callback_data\fR function. The \fIproc\fR argument
is a class providing information about the managed executable, like its
command-line and about the process in which it was executed.

.TP
\fB\-c\fR \fIscript-path\fR, \fB\-\-callbacks\fR \fIscript-path\fR
Specifies the filesystem path to a python script implementing one or more
of the driver callback functions described above.
If this option is not specified, then default, trivial callback functions
are used.

.SH "EXAMPLES"
.EX
.B atmost_driver -c ./mycallbacks.py
.EE
.PP
Starts the driver, using \fI./mycallbacks.py\fB as the callback script.

.SH "SEE ALSO"
.BR atmost (1),

.SH "AUTHOR"
Matus Chochlik, chochlik@gmail.com
.SH "COPYRIGHT"
Copyright (c) 2019 Matus Chochlik
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the Boost Software License, Version 1.0.
(See a copy at http://www.boost.org/LICENSE_1_0.txt)


atmost/atmost_driver.py
#!/usr/bin/env python
# coding: UTF-8
#  Copyright (c) 2019 Matus Chochlik

import os
import imp
import time
import json
import string
import errno
import socket
import signal
import psutil
import hashlib
import argparse
import multiprocessing

try:
    import selectors
except ImportError:
    import selectors2 as selectors

# ------------------------------------------------------------------------------
class AtmostCallbacks(object):
    # --------------------------------------------------------------------------
    def __init__(self):
        self.load_callback_data = lambda: None
        self.save_callback_data = lambda data: None

        self.process_initialized = lambda data, info: None
        self.let_process_go =\
            lambda data, info: len(info.active()) < multiprocessing.cpu_count()
        self.process_finished = lambda data, info: None
# ------------------------------------------------------------------------------
class AtmostDriverArgumentParser(argparse.ArgumentParser):
    # --------------------------------------------------------------------------
    def _try_import_callbacks_py(self, callbacks_path):
        callbacks = AtmostCallbacks()
        try:
            imp.acquire_lock()
            callbacks_py = imp.load_source(
                "atmost",
                os.path.join(callbacks_path)
            )
            try: callbacks.load_callback_data = callbacks_py.load_callback_data
            except: pass
            try: callbacks.save_callback_data = callbacks_py.save_callback_data
            except: pass
            try: callbacks.process_initialized = callbacks_py.process_initialized
            except: pass
            try: callbacks.let_process_go = callbacks_py.let_process_go
            except: pass
            try: callbacks.process_finished = callbacks_py.process_finished
            except: pass
        finally:
            imp.release_lock()
        return callbacks

    # --------------------------------------------------------------------------
    def __init__(self, **kw):
        argparse.ArgumentParser.__init__(self, **kw)

        self.add_argument(
            '-s', '--socket',
            metavar='SOCKET-FILE',
            dest='socket_path',
            nargs='?',
            type=os.path.realpath,
            default="/tmp/atmost.socket"
        )

        self.add_argument(
            '-c', '--callbacks',
            metavar='CALLBACK-SCRIPT',
            dest='callbacks_script',
            nargs='?',
            type=os.path.realpath,
            default=None
        )

        self.add_argument(
            '-u', '--update',
            metavar='INTERVAL',
            dest='update_interval',
            nargs='?',
            type=float,
            default=1
        )

    # --------------------------------------------------------------------------
    def process_parsed_options(self, options):
        options.callbacks = AtmostCallbacks()

        if options.callbacks_script is not None:
            path = options.callbacks_script
            if os.path.isfile(path):
                try:
                    options.callbacks = self._try_import_callbacks_py(path)
                except Exception as error:
                    self.error("failed to import callback script '%s': %s" % (
                        path,
                        str(error)
                    ))
            else:
                self.error("'%s' is not a file" % path)
        else:
            path = os.path.realpath("atmost_callbacks.py")
            if os.path.isfile(path):
                try:
                    options.callbacks = self._try_import_callbacks_py(path)
                except: pass
        #
        return options

    # --------------------------------------------------------------------------
    def parse_args(self):
        return self.process_parsed_options(
            argparse.ArgumentParser.parse_args(self)
        )

# ------------------------------------------------------------------------------
def make_argparser():
	return AtmostDriverArgumentParser(
            prog="atmost_driver",
            description="""
            Driver/server for the atmost utility
		"""
	)

# ------------------------------------------------------------------------------
def open_socket(options):
    if options.socket_path:
        try: os.unlink(options.socket_path)
        except OSError as os_error:
            if os_error.errno != errno.ENOENT:
                raise
        uds = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        uds.bind(options.socket_path)
        uds.listen(100)
        uds.setblocking(False)
        return uds

# ------------------------------------------------------------------------------
class AtmostProcess(object):
    # --------------------------------------------------------------------------
    def __init__(self, uid, parent):
        self._uid = uid
        self._parent = parent
        self._input_buff = str()
        self._output_buff = str()
        self._json = None
        self._pid = -1
        self._ciw = None
        self._arguid = None
        self._psutil = None
        self._responded = False
        self._create_time = time.time()
        self._ready_time = None
        self._let_go_time = None
        self._max_mem_usage = -1
        self._callback_data = None
        self._digest_trans = string.maketrans('0123456789', 'ghijklmnop')

    # --------------------------------------------------------------------------
    def __add__(self, other):
        if type(other) == AtmostProcess:
            return AtmostProcessList([self, other])
        if type(other) == AtmostProcessList:
            return AtmostProcessList([self] + list(other))
        raise TypeError

    # --------------------------------------------------------------------------
    def __eq__(self, other):
        return self._pid != -1 and other._pid != -1 and self._pid == other._pid

    # --------------------------------------------------------------------------
    def __ne__(self, other):
        return not self == other

    # --------------------------------------------------------------------------
    def update(self):
        try:
            self._max_mem_usage = max(
                self._max_mem_usage,
                self._psutil.memory_info().rss
            )
        except: pass

    # --------------------------------------------------------------------------
    def life_time(self):
        return time.time() - self._create_time

    # --------------------------------------------------------------------------
    def wait_time(self):
        if self._ready_time is not None:
            if self._let_go_time is not None:
                return self._let_go_time - self._ready_time
            else:
                return time.time() - self._ready_time
        return 0.0

    # --------------------------------------------------------------------------
    def run_time(self):
        if self._let_go_time is not None:
            return time.time() - self._let_go_time
        return 0.0

    # --------------------------------------------------------------------------
    def got_response(self):
        return self._responded

    # --------------------------------------------------------------------------
    def command_uid(self):
        return self._arguid

    # --------------------------------------------------------------------------
    def args(self):
        try: return self._json["args"]
        except: return []

    # --------------------------------------------------------------------------
    def exe(self):
        try: return self._json["args"][0]
        except: pass

    # --------------------------------------------------------------------------
    def basename(self):
        try: return os.path.basename(self._json["args"][0])
        except: pass

    # --------------------------------------------------------------------------
    def environment(self):
        try: return self._psutil.environ()
        except: pass

    # --------------------------------------------------------------------------
    def cwd(self):
        try: return self._psutil.cwd()
        except: return self._cwd

    # --------------------------------------------------------------------------
    def username(self):
        try: return self._psutil.username()
        except: pass

    # --------------------------------------------------------------------------
    def usernames(self):
        try: return [self._psutil.username()]
        except: []

    # --------------------------------------------------------------------------
    def status(self):
        try: return self._psutil.status()
        except: pass

    # --------------------------------------------------------------------------
    def num_threads(self):
        try: return self._psutil.num_threads()
        except: return 0

    # --------------------------------------------------------------------------
    def cpu_percent(self, interval=1):
        try: return self._psutil.cpu_percent(interval)
        except: return 0

    # --------------------------------------------------------------------------
    def memory_percent(self):
        try: return self._psutil.memory_percent()
        except: return 0

    # --------------------------------------------------------------------------
    def max_memory_bytes(self):
        return self._max_mem_usage

    # --------------------------------------------------------------------------
    def is_ready(self):
        return self._pid is not None and self._pid > 0

    # --------------------------------------------------------------------------
    def is_running(self):
        if self._pid > 0:
            try:
                os.kill(self._pid, 0)
                return True
            except OSError:
                pass
        return False

    # --------------------------------------------------------------------------
    def go(self, conn):
        self.update()
        self._responded = True
        self._let_go_time = time.time()
        self._output_buff = "OK-GO:%d\n" % self._pid
        self.handle_write(conn)

    # --------------------------------------------------------------------------
    def disconnect(self, conn):
        self._parent.disconnect_client(self._uid, conn)

    # --------------------------------------------------------------------------
    def string_list_hash(self, strings):
        h = hashlib.sha1()
        for s in strings:
            h.update(s)
        return h.hexdigest().translate(self._digest_trans)

    # --------------------------------------------------------------------------
    def command_line_uid(self):
        return self.string_list_hash(self.args())

    # --------------------------------------------------------------------------
    def set_callback_data(self, data):
        self._callback_data = data

    # --------------------------------------------------------------------------
    def callback_data(self):
        return self._callback_data

    # --------------------------------------------------------------------------
    def handle_read(self, conn):
        try:
            data = conn.recv(1024)
            if data:
                self._input_buff += str(data)
                try:
                    self._json = json.loads(self._input_buff)
                    self._pid = int(self._json["pid"])
                    self._psutil = psutil.Process(self._pid)
                    self._cwd = self._psutil.cwd()
                    self._arguid = self.command_line_uid()
                    self._parent.client_ready(self)
                    self._ready_time = time.time()
                except:
                    pass
            else:
                self.disconnect(conn)
        except socket.error:
                self.disconnect(conn)

    # --------------------------------------------------------------------------
    def handle_write(self, conn):
        if self._output_buff:
            done = conn.send(self._output_buff)
            self._output_buff = self._output_buff[done:]
            if not self._output_buff:
                self.disconnect(conn)

# ------------------------------------------------------------------------------
class AtmostProcessList(list):
    # --------------------------------------------------------------------------
    def __init__(self, items):
        list.__init__(self)
        self.extend(items)

    # --------------------------------------------------------------------------
    def __add__(self, other):
        if type(other) == AtmostProcessList:
            return AtmostProcessList(list(self) + list(other))
        if type(other) == AtmostProcess:
            return AtmostProcessList(list(self) + [other])
        raise TypeError

    # --------------------------------------------------------------------------
    def __len__(self):
        return list.__len__(self)

    # --------------------------------------------------------------------------
    def __iter__(self):
        return list.__iter__(self)

    # --------------------------------------------------------------------------
    def __getitem__(self, key):
        return list.__getitem__(self, key)

    # --------------------------------------------------------------------------
    def __delitem__(self, key):
        return list.__delitem__(self, key)

    # --------------------------------------------------------------------------
    def usernames(self):
        return list(set(n for p in self for n in p.usernames()))

    # --------------------------------------------------------------------------
    def num_threads(self):
        return sum([p.num_threads() for p in self])

    # --------------------------------------------------------------------------
    def cpu_percent(self, interval=1):
        return sum([p.cpu_percent(interval) for p in self])

    # --------------------------------------------------------------------------
    def memory_percent(self):
        return sum([p.memory_percent() for p in self])

    # --------------------------------------------------------------------------
    def max_memory_bytes(self):
        return sum([p.max_memory_bytes() for p in self])

# ------------------------------------------------------------------------------
class AtmostFilterContext(object):
    # --------------------------------------------------------------------------
    def __init__(self, clients, current):
        self._clients = clients
        self._current = current

    # --------------------------------------------------------------------------
    def total_memory(self):
        return psutil.virtual_memory().total

    # --------------------------------------------------------------------------
    def available_memory(self):
        return psutil.virtual_memory().available

    # --------------------------------------------------------------------------
    def current(self):
        return self._current

    # --------------------------------------------------------------------------
    def active(self):
        def _filter(x):
            return x.got_response() and x.is_running() and x != self._current
        return AtmostProcessList([x for x in self._clients if _filter(x)])

    # --------------------------------------------------------------------------
    def waiting(self):
        def _filter(x):
            return not x.got_response() and x.is_running() and x != self._current
        return AtmostProcessList([x for x in self._clients if _filter(x)])

# ------------------------------------------------------------------------------
class AtmostDriver(object):
    # --------------------------------------------------------------------------
    def __init__(self, selector, callbacks, callback_data):
        self._selector = selector
        self._callbacks = callbacks
        self._callback_data = callback_data
        self._clients = dict()

    # --------------------------------------------------------------------------
    def add_client(self, client_id, conn, client):
        self._clients[client_id] = (conn, client)
        self._selector.register(
            conn,
            selectors.EVENT_READ|
            selectors.EVENT_WRITE,
            data = client
        )

    # --------------------------------------------------------------------------
    def client_ready(self, client):
        self._callbacks.process_initialized(
            self._callback_data,
            client
        )

    # --------------------------------------------------------------------------
    def disconnect_client(self, client_id, conn):
        self._selector.unregister(conn)
        conn.close()
        del conn

    # --------------------------------------------------------------------------
    def remove_client(self, client_id, conn):
        self._callbacks.process_finished(
            self._callback_data,
            self._clients[client_id][1]
        )
        del conn
        del self._clients[client_id]

    # --------------------------------------------------------------------------
    def cleanup(self):
        to_be_removed = []
        for uid, (conn, client) in self._clients.items():
            if client.is_ready() and not client.is_running():
                to_be_removed.append((uid, conn))

        for uid, conn in to_be_removed:
            self.remove_client(uid, conn)

    # --------------------------------------------------------------------------
    def let_go(self, info):
        return self._callbacks.let_process_go(self._callback_data, info)

    # --------------------------------------------------------------------------
    def respond(self):
        clients = [client for conn, client in self._clients.values()]
        for (conn, curr) in self._clients.values():
            if curr.is_ready() and not curr.got_response():
                if self.let_go(AtmostFilterContext(clients, curr)):
                    curr.update()
                    curr.go(conn)

    # --------------------------------------------------------------------------
    def update(self):
        for conn, client in self._clients.values():
            client.update()
        self.cleanup()
        self.respond()

# ------------------------------------------------------------------------------
keep_running = True
# ------------------------------------------------------------------------------
def drive_atmost(options):
    callback_data = None
    try: callback_data = options.callbacks.load_callback_data()
    except: pass

    try:
        global keep_running
        lsock = open_socket(options)
        with selectors.DefaultSelector() as selector:
            driver = AtmostDriver(selector, options.callbacks, callback_data)
            selector.register(
                lsock,
                selectors.EVENT_READ,
                data = driver
            )

            client_id  = 0
            while keep_running:
                events = selector.select(timeout=options.update_interval)
                for key, mask in events:
                    if type(key.data) is AtmostDriver:
                        state = key.data
                        conn, addr = lsock.accept()
                        conn.setblocking(False)
                        client = AtmostProcess(client_id, state)
                        state.add_client(client_id, conn, client)
                        client_id += 1
                    elif type(key.data) is AtmostProcess:
                        client = key.data
                        conn = key.fileobj
                        if mask & selectors.EVENT_READ:
                            client.handle_read(conn)
                        if mask & selectors.EVENT_WRITE:
                            client.handle_write(conn)
                driver.update()
    finally:
        try:
            lsock.close()
            os.unlink(options.socket_path)
        except: pass

    try: options.callbacks.save_callback_data(callback_data)
    except: pass

# ------------------------------------------------------------------------------
def handle_interrupt(sig, frame):
    global keep_running
    keep_running = False
# ------------------------------------------------------------------------------
def main():
    signal.signal(signal.SIGINT, handle_interrupt)
    signal.signal(signal.SIGTERM, handle_interrupt)
    drive_atmost(make_argparser().parse_args())
    return 0
# ------------------------------------------------------------------------------
if __name__ == "__main__":
    exit(main())
# ------------------------------------------------------------------------------
